### PR TITLE
Don't suggest methods in certain cases

### LINKED ISSUE #43149: Rustc suggests .to_vec to convert a Vec<T> to Vec<U>
Hey all,

With the following piece of code ([rust playground](https://is.gd/yRqBjT)):
```rust
struct Foo {}
struct Bar {}

fn hey(pouet: Vec<Bar>){}

fn main() {
    let foo = vec![Foo {}];
    hey(foo);
}

```

rustc gives me the following error
```rust
error[E0308]: mismatched types
 --> a.rs:8:9
  |
8 |     hey(foo);
  |         ^^^ expected struct `Bar`, found struct `Foo`
  |
  = note: expected type `std::vec::Vec<Bar>`
             found type `std::vec::Vec<Foo>`
  = help: here are some functions which might fulfill your needs:
          - .to_vec()
```

which doesn't make sense since `.to_vec()` wouldn't in any case convert the inner type.

### LINKED ISSUE #43420: Incorrect compiler error message on mismatched types
I ran `cargo build` with some code that did not compile due to a type mismatch, error E0308. Help text was displayed, which was false/very misleading. 

Here's a reduced test case:
```
fn foo(b: &[u16]) {}

fn main() {
    let a: Vec<u8> = Vec::new();
    foo(&a);
}
```
After running `cargo build`, the error displayed was:
```
   Compiling test-compile v0.1.0 (file:///home/isaac/prog/rust/test-compile)
error[E0308]: mismatched types
 --> src/main.rs:5:9
  |
5 |     foo(&a);
  |         ^^ expected slice, found struct `std::vec::Vec`
  |
  = note: expected type `&[u16]`
             found type `&std::vec::Vec<u8>`
  = help: here are some functions which might fulfill your needs:
          - .as_slice()

error: aborting due to previous error(s)

error: Could not compile `test-compile`.
```
Focusing in on the line: expected slice, found struct `std::vec::Vec`

This is incorrect, because `std::vec::Vec` dereferences into slice. The actual type mismatch, which should have been displayed, is between `u16` and `u8`. I would expect the error to say:

expected `u16`, found `u8`

## Meta:

I discovered this on nightly, but reproduced it on stable.

`rustc --version --verbose`:

rustc 1.19.0 (0ade33941 2017-07-17)
binary: rustc
commit-hash: 0ade339411587887bf01bcfa2e9ae4414c8900d4
commit-date: 2017-07-17
host: x86_64-unknown-linux-gnu
release: 1.19.0
LLVM version: 4.0

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #43149.
Fixes #43420.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_typeck/check/demand.rs [hunk 1]
```diff
@@ -175,6 +178,61 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+
+    // For now it only checks slices, arrays and Vecs.
+    fn get_ty(&self, ty_: Ty<'tcx>) -> Option<Ty<'tcx>> {
+        match ty_.sty {
+            ty::TypeVariants::TyRawPtr(t) | ty::TypeVariants::TyRef(_, t) => self.get_ty(t.ty),
+            ty::TypeVariants::TyArray(ty_, _) | ty::TypeVariants::TySlice(ty_) => {
+                Some(ty_)
+            }
+            ty::TypeVariants::TyAdt(_, subs) => {
+                if ty_.to_string().split("::").last().unwrap_or("").starts_with("Vec<") {
```
Comment (arielb1) [inline src/librustc_typeck/check/demand.rs]: ?

Comment (GuillaumeGomez) [inline src/librustc_typeck/check/demand.rs]: Yep... I needed to check if it was a `Vec` but if you have a better to check it, I'm all for it!

Comment (arielb1) [inline src/librustc_typeck/check/demand.rs]: Like, ..., use an attribute?

**Comment by GuillaumeGomez** [comment]
@arielb1 will be working on a way better fix so no need to keep this one closed.

**Pull request closed without merging**